File: packages/swiftui/Sources/CocsoUI/CCProgress.swift
import SwiftUI

/// A determinate progress bar.
public struct CCProgress: View {
    private let value: Double
    private let total: Double
    private let variant: CCProgressVariant
    private let size: CCProgressSize
    private let label: String

    @Environment(\.colorScheme) private var colorScheme
    // The brand the app set at its root. Tokens and resolvers take it so a
    // design-system view draws the same primary the app does.
    @Environment(\.cocsoBrand) private var brand
    @Environment(\.accessibilityReduceMotion) private var reduceMotion

    public init(
        value: Double,
        total: Double = 100,
        label: String = "Progress",
        variant: CCProgressVariant = .primary,
        size: CCProgressSize = .md
    ) {
        self.value = value
        self.total = total
        self.label = label
        self.variant = variant
        self.size = size
    }

    public var body: some View {
        let style = CCProgressStyle.resolve(variant: variant, size: size, scheme: colorScheme, brand: brand)
        let fraction = total > 0 ? min(max(value / total, 0), 1) : 0
        let height = style.height ?? 8
        GeometryReader { geometry in
            // The recipe gives a radius per size. A capsule here ignored it,
            // and Android did not — the same bar was a different shape on the
            // two platforms.
            let shape = RoundedRectangle(cornerRadius: style.borderRadius ?? height / 2)
            ZStack(alignment: .leading) {
                shape.fill(style.bgColor ?? CocsoTokens.Color.surfaceNeutral(colorScheme, brand: brand))
                shape
                    .fill(style.fillColor ?? CocsoTokens.Color.interactivePrimary(colorScheme, brand: brand))
                    .frame(width: geometry.size.width * fraction)
                    // The web's `transition: width normal soft`: the bar fills
                    // to the new value rather than jumping to it.
                    .animation(CCMotion.fill(reduced: reduceMotion), value: fraction)
            }
        }
        .frame(height: height)
        // `role="progressbar"` is invalid without a name on the web, and the
        // equivalent here is a value with something naming it.
        .accessibilityElement()
        .accessibilityLabel(label)
        .accessibilityValue(Text("\(Int(fraction * 100))%"))
    }
}

File: packages/compose/src/main/kotlin/ai/cocso/ui/CCProgress.kt
package ai.cocso.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.progressBarRangeInfo
import androidx.compose.ui.semantics.ProgressBarRangeInfo
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp

/** A determinate progress bar. */
@Composable
fun CCProgress(
    value: Float,
    modifier: Modifier = Modifier,
    total: Float = 100f,
    label: String = "Progress",
    variant: CCProgressVariant = CCProgressVariant.primary,
    size: CCProgressSize = CCProgressSize.md,
) {
    val style = cCProgressStyle(variant = variant, size = size)
    val fraction = if (total > 0f) (value / total).coerceIn(0f, 1f) else 0f
    val height = style.height ?: 8.dp
    val shape = RoundedCornerShape(style.borderRadius ?: height / 2)

    Box(
        modifier = modifier
            .fillMaxWidth()
            .height(height)
            .clip(shape)
            .background(style.bgColor ?: CocsoTokens.Color.surfaceNeutral())
            // `role="progressbar"` is invalid without a name on the web, and the
            // equivalent here is a value with something naming it.
            .semantics {
                contentDescription = label
                progressBarRangeInfo = ProgressBarRangeInfo(fraction, 0f..1f)
            },
        contentAlignment = Alignment.CenterStart,
    ) {
        Box(
            modifier = Modifier
                .fillMaxWidth(fraction)
                .fillMaxHeight()
                .clip(shape)
                .background(style.fillColor ?: CocsoTokens.Color.interactivePrimary())
        )
    }
}
